core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java

```diff
@@ -81,13 +81,14 @@ public Schema record(Schema record, List<String> names, Iterable<Schema.Field> s
     List<Types.NestedField> expectedFields = struct.fields();
     for (int i = 0; i < expectedFields.size(); i += 1) {
       Types.NestedField field = expectedFields.get(i);
+      String sanitizedFieldName = AvroSchemaUtil.sanitize(field.name());
```

Comment:
When converting an Iceberg schema to an Avro schema if ever a field was sanitized, the original field name was stored as a field property under `AvroSchemaUtil.ICEBERG_FIELD_NAME_PROP` . Should we use that original name if it was set, instead of sanitizing all Iceberg fields before comparison.

Reply:
I tried your suggestion by getting the original name of the `field` by using `field.getProp(AvroSchemaUtil.ICEBERG_FIELD_NAME_PROP)` — the `expectedField` doesn't have it. Method `getProp` return null if the field has not been sanitized as the `ICEBERG_FIELD_NAME_PROP` is not set. This leads to more checks in all conditions bellow.

First, I guess we will still have to check if the `name`s are different for the cases where no sanitization is applied.
Second, we will have to check if the `ICEBERG_FIELD_NAME_PROP` is null itself.

Now, if the `AvroSchemaUtil.sanitize` is a pure function and the sanitization cases are correctly defined and desired to be applied everywhere  a field is `!validAvroName`, there should not be any issues using it in other parts over original fields — the output should be the same.

What do you think? Should I change this even though the code will be less readable due to checks?